core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -1040,6 +1040,11 @@ void testDyadicCollateOperator() {
         .columnType("VARCHAR(1) NOT NULL");
     expr("substring('a', 1, '3')")
         .columnType("VARCHAR(1) NOT NULL");
+
+    // Correctly processed null string and params.
```

Comment:
I suggest including a test for the 2-arg form too. Like:

```
expr("SUBSTRING('text' FROM NULL)").ok();
```

Reply:
done